internal/input/report.go

```diff
@@ -0,0 +1,207 @@
+	Successes int    `json:"successes"`
+	Failures  int    `json:"failures"`
+	Warnings  int    `json:"warnings"`
+	Result    string `json:"result"`
+	Note      string `json:"note,omitempty"`
+}
+
+// Possible formats the report can be written as.
+const (
+	SummaryMarkdown = "summary-markdown"
+	JSON            = "json"
+	YAML            = "yaml"
+	Summary         = "summary"
+	JUNIT           = "junit"
+)
+
+// WriteReport returns a new instance of Report representing the state of
+// the filepaths provided.
+func NewReport(inputs []Input, policy policy.Policy, data any, policyInput [][]byte) (Report, error) {
+	success := true
+
+	// Set the report success, remains true if all the files were successfully validated
+	for _, fpath := range inputs {
+		if !fpath.Success {
+			success = false
+			break
+		}
+	}
+
+	info, _ := version.ComputeInfo()
+
+	return Report{
+		Success:       success,
+		created:       time.Now().UTC(),
+		FilePaths:     inputs,
+		Policy:        policy.Spec(),
+		EcVersion:     info.Version,
+		Data:          data,
+		EffectiveTime: policy.EffectiveTime().UTC(),
+		PolicyInput:   policyInput,
+	}, nil
+}
+
+// WriteAll writes the report to all the given targets.
+func (r Report) WriteAll(targets []string, p format.TargetParser) (allErrors error) {
+	if len(targets) == 0 {
+		targets = append(targets, JSON)
+	}
+	for _, targetName := range targets {
+		target := p.Parse(targetName)
+
+		data, err := r.toFormat(target.Format)
+		if err != nil {
+			allErrors = multierror.Append(allErrors, err)
+		}
+
+		if !bytes.HasSuffix(data, []byte{'\n'}) {
+			data = append(data, "\n"...)
+		}
+
+		if _, err := target.Write(data); err != nil {
+			allErrors = multierror.Append(allErrors, err)
+		}
+	}
+	return
+}
+
+// toFormat converts the report into the given format.
+func (r *Report) toFormat(format string) (data []byte, err error) {
+	switch format {
+	case JSON:
+		data, err = json.Marshal(r)
+	case YAML:
+		data, err = yaml.Marshal(r)
+	case Summary:
+		data, err = json.Marshal(r.toSummary())
+	default:
+		return nil, fmt.Errorf("%q is not a valid report format", format)
+	}
+	return
+}
+
+// toSummary returns a condensed version of the report.
+func (r *Report) toSummary() summary {
+	pr := summary{}
+	for _, cmp := range r.FilePaths {
+		c := inputSummary{
+			FilePath:        cmp.FilePath,
+			TotalViolations: len(cmp.Violations),
+			TotalWarnings:   len(cmp.Warnings),
+
+			// Because cmp.Successes does not get populated unless the --show-successes
+			// flag was set, cmp.SuccessCount is used here instead of len(cmp.Successes)
+			TotalSuccesses: cmp.SuccessCount,
+
+			Success:    cmp.Success,
+			Violations: condensedMsg(cmp.Violations),
+			Warnings:   condensedMsg(cmp.Warnings),
+			Successes:  condensedMsg(cmp.Successes),
+		}
+		pr.FilePaths = append(pr.FilePaths, c)
+	}
+	return pr
+}
+
+// condensedMsg reduces repetitive error messages.
+func condensedMsg(results []evaluator.Result) map[string][]string {
```

Comment:
There's some important logic here. We should really consider not duplicating it.

Reply:
I'm going to leave this for a later iteration.